content/blog/2017/2017-01-31-pipeline-scalability-best-practice.adoc

```diff
@@ -0,0 +1,197 @@
+---
+layout: post
+title: "Best Practices for Scalable Pipeline Code"
+tags:
+- pipeline
+- performance
+- scalability
+author: svanoort
+---
+====
+This by link:https://github.com/svanoort[Sam Van Oort], Software Engineer at
+link:http://cloudbees.com[CloudBees] and contributor to the Jenkins project.
+====
+
+=== Promises
+
+Today I'm going to show you best practices to write scalable and robust Jenkins pipeline. This is drawn from a
+combination of work with the internals of Pipeline and observations with high-scalability users.
+
+Pipeline code works beautifully for its intended role of orchestrating
+build/test/deploy/administer cycles.  As it is pressed into more complex roles and unanticipated uses, some users hit issues.
+In these cases, applying the best practices can make the difference between:
+
+* A single master running
+https://www.cloudbees.com/so-you-want-build-worlds-biggest-jenkins-cluster[hundreds
+of concurrent builds] on low end hardware (4 CPU cores and 4 GB of
+heap)
+* Running a couple dozen builds and bringing a master to its knees or
+crashing it...even with 16+ CPU cores and 20+ GB of heap!
+
+This has been seen in the wild.
+
+[[fundamentals]]
+Fundamentals
+------------
+
+To understand pipeline behavior you must understand a few points about
+how it executes.
+
+1.  All of the pipeline logic except for steps runs on the master (the
```

Comment:
Aggh @rtyler I spent probably 30 minutes agonizing over how to phrase that sentence and you had a better way. Thank you! :)  I ought to just let review sort it out.